packages/mattermost-redux/src/reducers/entities/channels.ts

```diff
@@ -898,7 +897,7 @@ function roles(state: RelationOneToOne<Channel, Set<string>> = {}, action: Gener
         const newRoles = splitRoles(channelMember.roles);
 
         // If roles didn't change no need to update state
-        if (shallowEquals(oldRoles, newRoles)) {
+        if (isEqual(oldRoles, newRoles)) {
```

Comment:
reason we added that before was to return early by just doing a shallow compare because of performance reason, lodash isEqual on the other hand is deep compare.

Reply:
While I agree we shouldn't opt-in to deep compare, in this case, `oldRoles` and `newRoles` are sets, not arrays. It seems to me that the library does not properly support them and gives a false positive, which breaks the intended functionality except if we should convert sets to arrays every time, which is probably less efficient either way. I tried it a bit to verify it again in [this codesandbox](https://codesandbox.io/embed/laughing-yalow-qg7ig8). Am I missing something?

Reply:
Thank you so much for attaching the codesandbox, it proves that shallow equal doesnt compare sets. I am not sure why we didnt caught that earlier.